src/magic_folder/snapshot.py

```diff
@@ -0,0 +1,172 @@
+        the grid.
+    :param HTTPClient treq: An ``HTTPCLient`` or similar object to use to
+        make the queries.
+    :return Deferred[unicode]: The readcap associated with the newly created
+        unlinked file.
+    """
+    node_url = DecodedURL.from_text(unicode(nodeurl, 'utf-8'))
+    url = node_url.child(
+        u"uri",
+    ).add(
+        u"format",
+        u"CHK",
+    )
+
+    # XXX: check whether we need to set any headers.
+    put_uri = url.to_uri().to_text().encode("ascii")
+    with open(str(filepath), "rb") as file:
+        data = file.read()
+
+    response = yield treq.put(put_uri, data)
+    if response.code == OK or response.code == CREATED:
+        result = yield readBody(response)
+        returnValue(result)
+    else:
+        raise Exception("Error response PUT {} - {}".format(put_uri, response))
+
+@inlineCallbacks
+def tahoe_create_snapshot_dir(nodeurl, content, parents, timestamp, treq):
+    """
+    :param unicode content: readcap for the content.
+    :param [unicode] parents: List of parent snapshot caps
+    :param unicode author: readcap that represents the author pubkey
+    :param integer timestamp: POSIX timestamp that represents the creation time
+    :return Deferred[unicode]: The readcap associated with the newly created
+        snapshot.
+    """
+
+    # dict that would be serialized to JSON
+    body = \
+    {
+        u"content": [ "filenode", { u"ro_uri": content,
+                                    u"metadata": { } } ],
+        u"version": [ "filenode", { u"ro_uri": str(SNAPSHOT_VERSION),
+                                    u"metadata": { } } ],
+        u"timestamp": [ "filenode", { u"ro_uri": str(timestamp),
+                                      u"metadata": { } } ],
+    }
+
+    # populate parents
+    # The goal is to populate the dictionary with keys u"parent0", u"parent1" ...
+    # with corresponding dirnode values that point to the parent URIs.
+    if parents != []:
+        parent_nodes =  [ [ "dirnode", { u"ro_uri": p } ] for p in parents ]
+        for (k, v) in zip(count(0), parent_nodes):
+            body[unicode("parent" + str(k), 'utf-8')] = v
+
+    body_json = json.dumps(body)
+
+    node_url = DecodedURL.from_text(unicode(nodeurl, 'utf-8'))
+
+    # POST /uri?t=mkdir-immutable
+    url = node_url.child(
+        u"uri",
+    ).add(
+        u"t",
+        u"mkdir-immutable"
+    )
+
+    post_uri = url.to_uri().to_text().encode("ascii")
+    response = yield treq.post(post_uri, body_json)
+    if response.code != OK:
+        returnValue((yield bad_response(url, response)))
+
+    result = yield readBody(response)
+    returnValue(result)
+
+@inlineCallbacks
+def _store_file_immutable(nodeurl, filepath):
+
+    from twisted.internet import reactor
+    treq = HTTPClient(Agent(reactor))
+    rocap = yield tahoe_put_immutable(nodeurl, filepath, treq)
+
+    returnValue(rocap)
+
+@inlineCallbacks
+def snapshot_create(node_directory, filepath, parents):
```

Comment:
At risk of this just being "bikeshed" territory ;) I think it would be more-consistent to be named `create_snapshot` to mirror other tahoe and magic-folder code. Based on a definitely-scientific `grep` survey, `def create_*` is more popular than `def *_create` in both codebases. See-also the comment earlier about "a file-like object" versus file-path.